README.rst
slackpdb - Remotely and collaboratively debug your Python application via a Slack channel
=========================================================================================

.. image:: https://travis-ci.org/coddingtonbear/slackpdb.svg?branch=master
    :target: https://travis-ci.org/coddingtonbear/slackpdb

Slackpdb is an adaptation of rpdb that, instead of opening a port and
allowing you to debug over telnet, connects to a configurable Slack
channel so you can collaboratively debug an application remotely.

.. code-block::

    import slackpdb
    slackpdb.set_trace(
        token="<your slack token>",
        channel="#debugger_hangout",
    )

Upon reaching ``set_trace()``, your script will "hang" and the only way to get
it to continue is to access slackpdb by talking to the user that connected to the
above IRC channel.

To interact with the debugger, just send messages in the channel prefixed
with the username it announces itself as, or simply "!".

For example, the following two commands are equivalent, and each will
display the pdb help screen (be sure to replace 'MyHostname' with whatever
username the bot selected)::

    !help

::

    slackpdb: help

Installation
------------

From ``pip``::

    pip install slackpdb

Options
-------

* ``token``: A token to use for connecting to Slack; if you do not have one,
  you can quickly generate one at https://my.slack.com/services/new/bot.  Note
  that if this is not specified, it may be gathered from the
  ``SLACKPDB_API_TOKEN`` environment variable, or, if the ``django`` option is
  set to ``True``, from a Django setting named ``SLACKPDB_API_TOKEN``.
* ``channel`` (**REQUIRED**): The name of the channel (starting with ``#``).
* ``limit_access_to``: A list of nicknames that
  are allowed to interact with the debugger.
* ``activation_timeout``: Wait maximally this number of seconds for
  somebody to interact with the debugger in the channel before
  disconnecting and continuing execution.  Default: ``60`` seconds.
* ``django``: (Default: ``False``) Attempt to gather Slack API token from
  a django setting named ``SLACKPDB_API_TOKEN`` if otherwise unspecified.

Security Disclaimer
-------------------

The way that this library works is **inherently** **dangerous**; given that
you're able to execute arbitrary Python code from within your debugger,
it is strongly recommended that you take all reasonable measures to ensure
that you control who are able to execute debugger commands.

Just to make absolutely sure this is clear: you're both responsible for
determining what level of risk you are comfortable with, and for taking
appropriate actions to mitigate that risk.

As is clearly and thunderously stated library's license (see the included
``LICENSE.txt``)::

    THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
    ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
    IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
    ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
    FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
    DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
    OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
    HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
    LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
    OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
    SUCH DAMAGE.

Good luck, and happy debugging!

Troubleshooting
---------------

If you do not see the bot entering your specified channel, try increasing
the logging level by adding the following lines above your trace to gather
a little more information about problems that may have occurred while 
connecting to the IRC server:

.. code-block::

   import logging
   logging.basicConfig(filename='/path/to/somewhere.log', level=logging.DEBUG)

Author(s)
---------
Adam Coddington <me@adamcoddington.net> - http://adamcoddington.net/

This library is a fork of rpdb, and the underpinnings of this library
are owed to Bertrand Janin <b@janin.com> - http://tamentis.com/ and
all other contributors to `rpdb <https://github.com/tamentis/rpdb>`
including the following:

 - Ken Manheimer - @kenmanheimer
 - Steven Willis - @onlynone
 - Jorge Niedbalski R <niedbalski@gmail.com>
 - Cyprien Le Pannérer <clepannerer@edd.fr>
 - k4ml <kamal.mustafa@gmail.com>
 - Sean M. Collins <sean@coreitpro.com>


CHANGES.txt
2.0 (2017-06-07)
================

 - Forked form ircpdb!

1.8 (2016-08-19)
================

 - Switched to Gist instead of Dpaste.
 - Added functionality allowing one to specify their own Paste backend.

1.7 (2015-01-28)
================

 - Added 'activation_timeout' timer that defines a time limit for the first
   interaction with the IRC channel before disconnecting and continuing
   execution.

1.6 (2014-12-20)
================

 - Adds a django template tag {% set_trace %}; this is usable for opening
   up an ircpdb channel during template parsing.

1.5 (2014-12-04)
================

 - Adds functionality for specifying configuration parameters via a single
   argument containing a URI.  Rather than specifying potentially a dozen
   separate parameters, one can now provide a single string; for example:
   "irc+ssl://bot@chat.freenode.net:6667/#mychannel?limit_access_to=me".
 - Adds ability to specify default configuration parameters via environment
   variable named 'DEFAULT_IRCPDB_URI'.

1.4 (2014-11-11)
================

 - Simpler internal handling of prompt.
 - Adding internal support for IRC CTCP commands.
 - Pinning jaraco/irc library version at 0.9 to preserve Python 2.6 support.

1.3 (2014-11-01)
================

 - Reduced security concerns by making it impossible to configure the
   debugger to respond to all usernames.
 - Fixed a UI glitch in which the prompt would be displayed in inappropriate
   situations.

1.2 (2014-10-31)
================

 - Many UI improvements including:
   - Display a prompt after the response has been written.
 - Added '!!forbid' command.
 - Added a thunderous warning to the readme about the security implications
   of this library.

1.1 (2014-10-30)
================

 - Added functionality for automatically sending long messages to Dpaste.
 - Added additional commands including `!!set_dpaste_minimum_response_length`
   and `!!set_message_wait_seconds`.
 - Send a hello message once connected to the channel.

1.0 (2014-10-29)
==================

 - Forked from rpdb to, instead of opening a socket to interact with,
   opening a connection to an IRC channel.
   
0.1.5 (2014-10-16)
==================

 - Write addr/port to stderr instead of stdout (thanks to @onlynone).
 - Allow for dynamic host port (thanks to @onlynone).
 - Make q/quit do proper cleanup (@kenmanheimer)
 - Benignly disregard repeated rpdb.set_trace() to same port as currently
   active session (@kenmanheimer)
 - Extend backwards compatibility down to Python 2.5 (@kenmanheimer)

0.1.4 (2014-04-28)
==================

 - Expose the addr, port arguments to the set_trace method (thanks to @niedbalski).

0.1.3 (2013-08-02)
==================

 - Remove a try/finally that seemed to shift the trace location (thanks to k4ml@github).

0.1.2 (2012-01-26)
==================

 - Catch IOError raised by print in initialization, it may not work in some
   environments (e.g. mod_wsgi). (Menno Smits)

0.1.1 (2010-05-09)
==================

Initial release.



slackpdb/bot.py
import fcntl
import logging
from multiprocessing import Queue
import os
import random
import socket
import textwrap
import time

import six
from slackclient import SlackClient

from .exceptions import SlackpdbError


logger = logging.getLogger(__name__)


class SlackpdbBot(SlackClient):
    PROMPT = u'_Ready for command..._'

    _CHANNEL_INFO = {}
    _USER_INFO = {}

    def __init__(
        self, token, channel, limit_access_to, activation_timeout,
        **connect_params
    ):
        self.token = token
        self.channel = channel
        self.queue = Queue()
        self.activated = False
        self.activation_timeout = activation_timeout
        self.limit_access_to = limit_access_to

        super(SlackpdbBot, self).__init__(token)

        self.token_info = self.get_token_info()
        self.channel_info = self.get_debugger_channel_info()

    def send_welcome(self):
        self.send_channel_message(
            [
                u"*Debugger ready on host %s*" % socket.gethostname(),
                (
                    (
                        u"_The following users are able to interact with this "
                        u"debugger: %s._" % (
                            ', '.join(self.limit_access_to)
                        )
                    ) if self.limit_access_to else (
                        u"_All users in this channel can interact with this "
                        u"debugger._"
                    )
                ),
                (
                    u"_Please prefix debugger commands with '@{username}'. "
                    u"For pdb help, say '@{username} help'; for a list of "
                    u"slackpdb-specific commands, say '@{username} "
                    u"!help'._".format(
                        username=self.token_info['user'],
                    )
                )
            ],
            as_snippet=False,
        )

    def get_token_info(self):
        result = self.api_call('auth.test')

        if not result.get('ok'):
            raise ValueError('Unable to find username!')

        return result

    def get_debugger_channel_info(self):
        channels = self.api_call('channels.list')['channels']

        for channel in channels:
            if channel.get('name') == self.channel.strip('#'):
                return self.get_channel_info(channel.get('id'))

        raise ValueError('Specified channel does not exist!')

    def get_user_info(self, user):
        if not user:
            return {}
        if user not in self._USER_INFO:
            result = self.api_call(
                'users.info',
                user=user
            )

            if result.get('ok'):
                self._USER_INFO[user] = result['user']

        return self._USER_INFO.get(user, {})

    def get_channel_info(self, channel):
        if not channel:
            return {}
        if channel not in self._CHANNEL_INFO:
            result = self.api_call(
                'channels.info',
                channel=channel
            )

            if result.get('ok'):
                self._CHANNEL_INFO[channel] = result['channel']

        return self._CHANNEL_INFO.get(channel, {})

    def message_should_be_processed(self, message):
        if message.get('type') != 'message':
            return False

        if message.get('channel') == 'C4MN906RY':
            import pdb
            pdb.set_trace()

        if message.get('edited'):
            logger.warning(
                "Ignoring edited message: %s",
                message,
            )
            return False

        channel = self.get_channel_info(message.get('channel'))

        if channel.get('name') != self.channel.strip('#'):
            return False

        user = self.get_user_info(message.get('user'))
        if not channel or not user:
            return False

        if message.get('user') == self.token_info['user_id']:
            return False

        if not channel.get('is_channel', False):
            return False

        try:
            prefix, message = (
                message.get('text', '').split(' ', 1)
            )
        except ValueError:
            return False

        if (
            prefix.strip(':').strip('<').strip('>').strip('@')
            not in (
                self.token_info['user'],
                self.token_info['user_id'],
            )
        ):
            return False

        if (
            self.limit_access_to
            and user.get('name') not in self.limit_access_to
        ):
            return False

        return True

    def send_user_message(self, nickname, message):
        if not isinstance(message, six.text_type):
            message = u'\n'.join(message)

        logger.debug(
            "Sending user '%s' message '%s'.",
            nickname,
            message,
        )

        result = self.api_call(
            'chat.postMessage',
            channel=nickname,
            text=message,
            username=self.token_info['user'],
            as_user=False,
        )

        logger.debug(
            "Result: %s",
            result,
        )

    def send_channel_message(self, message, as_snippet=True):
        if not isinstance(message, six.text_type):
            message = u'\n'.join(message)

        logger.debug(
            "Sending channel '%s' %s.",
            message,
            'as snippet' if as_snippet else 'as message'
        )

        if as_snippet:
            result = self.api_call(
                'files.upload',
                channels=self.channel,
                filetype='python',
                content=message,
            )
        else:
            result = self.api_call(
                'chat.postMessage',
                channel=self.channel,
                text=message,
                username=self.token_info['user'],
                as_user=False,
            )

        logger.debug(
            "Result: %s",
            result,
        )

    def do_command(self, message, ):
        self.activated = True

        _, cmd = message.get('text').split(' ', 1)
        user = self.get_user_info(message.get('user'))

        logger.debug('Received command: %s', message)

        nickname = user.get('name')

        if cmd.startswith("!allow"):
            allows = cmd.split(' ')
            usernames = allows[1:]
            if not self.limit_access_to:
                self.limit_access_to.append(nickname)
            self.limit_access_to.extend(usernames)
            self.send_channel_message(
                u"The following users have been granted access to the "
                u"debugger: %s." % (
                    ', '.join(usernames)
                )
            )
        elif cmd.startswith("!forbid"):
            forbids = cmd.split(' ')
            usernames = forbids[1:]
            try:
                for username in usernames:
                    self.limit_access_to.remove(username)
                self.send_channel_message(
                    u"The following users have been forbidden access to the "
                    u"debugger: %s." % (
                        ', '.join(usernames)
                    )
                )
            except ValueError:
                self.send_channel_message(
                    u"The users %s are not in the 'allows' list.  You must "
                    u"have a defined 'allows' list to remove users from "
                    u"it." % (
                        ', '.join(usernames)
                    )
                )
            if not self.limit_access_to:
                self.send_channel_message(
                    u"All users are allowed to interact with the debugger. "
                )
        elif cmd.startswith("!help"):
            available_commands = textwrap.dedent(u"""
                Available Commands:
                * @{username} allow NICKNAME
                  Add NICKNAME to the list of users that are allowed to
                  interact with the debugger. Current value: {limit_access_to}.

                * @{username} forbid NICKNAME
                  Remove NICKNAME from the list of users that are allowed
                  to interact with the debugger.
            """.format(
                username=self.token_info['user'],
                limit_access_to=self.limit_access_to,
            )).strip()
            self.send_channel_message(
                available_commands,
            )
        else:
            logger.debug(
                'Sending command `%s` to debugger for message %s',
                cmd.strip(),
                message,
            )
            self.queue.put(cmd.strip())

    def send_prompt(self):
        self.send_channel_message(self.PROMPT, as_snippet=False)

    def api_call(self, *args, **kwargs):
        logger.debug('Sending API command: %s %s', args, kwargs)
        result = super(SlackpdbBot, self).api_call(*args, **kwargs)
        logger.debug('Received API result: %s', result)

        return result

    def process_forever(self, inhandle, outhandle, timeout=0.1):
        # Let's mark out inhandle as non-blocking
        fcntl.fcntl(inhandle, fcntl.F_SETFL, os.O_NONBLOCK)
        # Used for keeping track of when the bot was started
        # so we can disconnect for inactivity.
        started = time.time()
        # Keeps track of whether or not we're in the process of
        # disconnecting (queued the 'q' command)
        disconnecting = False

        if not self.rtm_connect():
            raise SlackpdbError("Unable to connect to Slack RTM service.")

        self.send_welcome()

        if self.token_info['user_id'] not in self.channel_info['members']:
            self.send_channel_message(
                '*The debugger is not currently invited to this channel! '
                'The debugger bot must be in this channel for you to interact '
                'with it; please invite it by saying `/invite @%s`.*' % (
                    self.token_info['user']
                ),
                as_snippet=False,
            )

        while True:
            try:
                messages = inhandle.read()
            except IOError:
                messages = None
            if messages:
                for message in messages.split('(Pdb)'):
                    stripped = message.strip()
                    if stripped:
                        logger.debug('>> %s', stripped)
                        self.send_channel_message(stripped.decode('utf8'))
                self.send_prompt()

            messages = self.rtm_read()
            for message in messages:
                logger.debug('Incoming RTM: %s', message)
                if self.message_should_be_processed(message):
                    self.do_command(message)

            while True:
                if self.queue.empty():
                    break
                message = self.queue.get(block=False)
                logger.debug('<< %s', message)
                outhandle.write(u'%s\n' % message)
                outhandle.flush()

            if (
                time.time() > started + self.activation_timeout and
                not self.activated and
                not disconnecting
            ):
                self.send_channel_message(
                    u"No response received within %s seconds; "
                    u"disconnecting due to inactivity." % (
                        self.activation_timeout
                    ),
                    as_snippet=False,
                )
                disconnecting = True
                self.queue.put('c')


slackpdb/debugger.py
import logging
import os
import pdb
import socket
import sys
from threading import Thread
import traceback

import six

from .bot import SlackpdbBot


logger = logging.getLogger(__name__)


DEFAULT_PARAMS = {
    'channel': '#debugger',
}


class Slackpdb(pdb.Pdb):
    def __init__(self, token, **kwargs):
        """Initialize the socket and initialize pdb."""
        params = DEFAULT_PARAMS.copy()
        params.update(kwargs)

        # Backup stdin and stdout before replacing them by the socket handle
        self.old_stdout = sys.stdout
        self.old_stdin = sys.stdin
        self.read_timeout = 0.1

        if not token:
            raise ValueError("A slack API token must be provided.")
        if not params.get('limit_access_to'):
            logger.warning(
                "No nickname limitations set!  Any user talking to the "
                "debugger user can execute arbitrary code on the host "
                "machine!"
            )
        elif isinstance(params.get('limit_access_to'), six.string_types):
            params['limit_access_to'] = [params.get('limit_access_to')]

        connect_params = {}
        if not params.get('channel'):
            raise ValueError(
                "You must specify a channel to connect to using the "
                "`channel` keyword argument."
            )

        # Writes to stdout are forbidden in mod_wsgi environments
        try:
            logger.info(
                "slackpdb has connected to slack in %s\n",
                params.get('channel')
            )
        except IOError:
            pass

        r_pipe, w_pipe = os.pipe()
        # The A pipe is from the bot to pdb
        self.p_A_pipe = os.fdopen(r_pipe, 'r')
        self.b_A_pipe = os.fdopen(w_pipe, 'w')

        r_pipe, w_pipe = os.pipe()
        # The B pipe is from pdb to the bot
        self.b_B_pipe = os.fdopen(r_pipe, 'r')
        self.p_B_pipe = os.fdopen(w_pipe, 'w')

        pdb.Pdb.__init__(
            self,
            stdin=self.p_A_pipe,
            stdout=self.p_B_pipe,
        )

        self.bot = SlackpdbBot(
            token=token,
            channel=params.get('channel'),
            limit_access_to=params.get('limit_access_to'),
            activation_timeout=params.get('activation_timeout', 60),
            **connect_params
        )

    def shutdown(self):
        """Revert stdin and stdout, close the socket."""
        sys.stdout = self.old_stdout
        sys.stdin = self.old_stdin
        pipes = [
            self.p_A_pipe,
            self.p_B_pipe,
            self.b_A_pipe,
            self.b_B_pipe
        ]
        for pipe in pipes:
            try:
                pipe.close()
            except IOError:
                logger.warning(
                    "IOError encountered while closing a pipe; messages "
                    "may have been lost."
                )

    def do_continue(self, arg):
        """Clean-up and do underlying continue."""
        try:
            return pdb.Pdb.do_continue(self, arg)
        finally:
            self.shutdown()

    do_c = do_cont = do_continue

    def do_quit(self, arg):
        """Clean-up and do underlying quit."""
        try:
            return pdb.Pdb.do_quit(self, arg)
        finally:
            self.shutdown()

    do_q = do_exit = do_quit


def set_trace(*args, **kwargs):
    """Wrapper function to keep the same import x; x.set_trace() interface.

    We catch all the possible exceptions from pdb and cleanup.

    """
    if not args and os.environ.get('SLACKPDB_API_TOKEN'):
        args = (os.environ.get('SLACKPDB_API_TOKEN'), )
    if not args and kwargs.get('django', False):
        from django.conf import settings
        args = (settings.SLACKPDB_API_TOKEN, )
        kwargs.pop('django')

    debugger = Slackpdb(*args, **kwargs)

    try:
        slack_feeder = Thread(
            target=debugger.bot.process_forever,
            args=(debugger.b_B_pipe, debugger.b_A_pipe, ),
        )
        slack_feeder.daemon = True
        slack_feeder.start()

        debugger.set_trace(sys._getframe().f_back)
    except Exception:
        debugger.shutdown()
        traceback.print_exc()
